### PR TITLE
Buffer API changes, Optimization and refactor of ResizeProcessors

### PR DESCRIPTION
**Pull request by antonfirsov**
### Prerequisites
- [x] I have written a descriptive pull-request title
- [x] I have verified that there are no overlapping [pull-requests](https://github.com/JimBobSquarePants/ImageSharp/pulls) open
- [x] I have verified that I am following matches the existing coding patterns and practise as demonstrated in the repository. These follow strict Stylecop rules :cop:.
- [x] I have provided test coverage for my change (where applicable)

## Buffer API changes
- Refactored `BufferPointer<T>` to `BufferSpan<T>`. It's API is very similar to `System.Span` now.
- Introduced `PinnedImageBuffer<T>`: a specialization of `PinnedBuffer<T>` that holds 2D area of data. It's very similar to `PixelAccessor<T>`, but the elements are not constrained to be `IPixel<T>`-s.
- Introduced an `IPinnedImageBuffer<T>` interface for common functionalities of `PixelAcccessor<T>` and `PinnedImageBuffer<T>`. (See `PinnedImageBufferExtensions`)

## ResizeProcessor changes
- Implemented first 2 points of #139
  - Weights data is stored in a contigous memory area now (see `WeightsWindow` and `WeightsBuffer`)
  - In the final implementation the execution time dropped about 30%
- Merged `ResizeProcessor` and `CompandResizeProcessor`
- Removed several code duplications

## Benchmarks results

**CPU specs:**
Processor=Intel(R) Core(TM) i7-4810MQ CPU 2.80GHz, ProcessorCount=8
(4 physical cores)

#### Before:
Method |       Mean |    StdDev | Scaled | Scaled-StdDev | Allocated |
---------------------------- |----------- |---------- |------- |-------------- |---------- |
'System.Drawing Resize' | 52.4481 ms | 0.7776 ms |   1.00 |          0.00 |     512 B |
'ImageSharp Resize' | 35.2544 ms | 1.1755 ms |   0.67 |          0.02 |     17 MB |
'ImageSharp Compand Resize' | 65.6250 ms | 1.3946 ms |   1.25 |          0.03 |     17 MB |

#### After:
Method |       Mean |    StdErr |    StdDev | Scaled | Scaled-StdDev | Allocated |
---------------------------- |----------- |---------- |---------- |------- |-------------- |---------- |
'System.Drawing Resize' | 51.4692 ms | 0.1478 ms | 0.5330 ms |   1.00 |          0.00 |     512 B |
'ImageSharp Resize' | 24.4749 ms | 0.3067 ms | 1.2647 ms |   0.48 |          0.02 |  17.02 MB |
'ImageSharp Compand Resize' | 43.8705 ms | 0.1408 ms | 0.5451 ms |   0.85 |          0.01 |  17.02 MB |

#### A fully sequential version of "after" on the same machine (no Parallel.For-s at all):
Method |        Mean |    StdDev |      Median | Scaled | Scaled-StdDev | Allocated |
---------------------------- |------------ |---------- |------------ |------- |-------------- |---------- |
'System.Drawing Resize' |  52.3989 ms | 1.3285 ms |  52.0098 ms |   1.00 |          0.00 |     512 B |
'ImageSharp Resize' |  58.7112 ms | 1.5875 ms |  57.6037 ms |   1.12 |          0.04 |   16.9 MB |
'ImageSharp Compand Resize' | 149.1699 ms | 0.1799 ms | 149.1820 ms |   2.85 |          0.07 |   16.9 MB |

**Conclusion:** considering that I have 4 physical cores, the `2.85/0.85 = 3.35` parallel speedup is quite good on the Compand stuff. It means, we should find an efficient way to keep the parallelism for non-server users.

### CODE REVIEW

**Comment by tocsoft** [comment]
I can't see anything thing in here that worries me. Looking good 👍

File: src/ImageSharp/Processing/Processors/Transforms/ResamplingWeightedProcessor.Weights.cs [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+            private PinnedImageBuffer<float> dataBuffer;
+
+            /// <summary>
+            /// Initializes a new instance of the <see cref="WeightsBuffer"/> class.
+            /// </summary>
+            /// <param name="sourceSize">The size of the source window</param>
+            /// <param name="destinationSize">The size of the destination window</param>
+            public WeightsBuffer(int sourceSize, int destinationSize)
+            {
+                this.dataBuffer = new PinnedImageBuffer<float>(sourceSize, destinationSize);
```
Comment (JimBobSquarePants) [inline src/ImageSharp/Processing/Processors/Transforms/ResamplingWeightedProcessor.Weights.cs]: Any reason for not using the static method `CreateClean` to generate the clean buffer?

Comment (antonfirsov) [inline src/ImageSharp/Processing/Processors/Transforms/ResamplingWeightedProcessor.Weights.cs]: No it was occasional. Could be more readable with CreateClean().

**Comment by antonfirsov** [comment]
I think it's enough to move out if-s from the inner loop. We can benchmark a version where they are outside the outer loops, but I don't think it's worth the efforts.

File: src/ImageSharp/Processing/Processors/Transforms/ResamplingWeightedProcessor.Weights.cs (hunk 1 above)
Comment (JimBobSquarePants) [inline src/ImageSharp/Processing/Processors/Transforms/ResamplingWeightedProcessor.Weights.cs]: Cool. I would say so. Intent is really obviously then.

**Comment by JimBobSquarePants** [comment]
I'll have a wee go just in case.

**Pull request merged**